src/Libraries/Microsoft.Extensions.AI/Embeddings/NormalizationEmbeddingGenerator.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#if NET // requires generic math interfaces
+using System.Collections.Generic;
+using System.Numerics;
+using System.Numerics.Tensors;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.AI;
+
+/// <summary>A delegating embedding generator that normalizes embedding vectors not already annotated as being normalized.</summary>
+/// <typeparam name="TInput">Specifies the type of the input passed to the generator.</typeparam>
+/// <typeparam name="TEmbeddingValue">Specifies the type of the values in the embedding vectors produced by the generator.</typeparam>
+public sealed class NormalizationEmbeddingGenerator<TInput, TEmbeddingValue> : DelegatingEmbeddingGenerator<TInput, Embedding<TEmbeddingValue>>
+    where TEmbeddingValue : IRootFunctions<TEmbeddingValue>
+{
+    /// <summary>Initializes a new instance of the <see cref="NormalizationEmbeddingGenerator{TInput, TEmbeddingElement}"/> class.</summary>
+    /// <param name="innerGenerator">The underlying <see cref="IEmbeddingGenerator{TInput, Embedding}"/>.</param>
+    public NormalizationEmbeddingGenerator(IEmbeddingGenerator<TInput, Embedding<TEmbeddingValue>> innerGenerator)
+        : base(innerGenerator)
+    {
+    }
+
+    /// <inheritdoc/>
+    public override async Task<GeneratedEmbeddings<Embedding<TEmbeddingValue>>> GenerateAsync(
+        IEnumerable<TInput> values, EmbeddingGenerationOptions? options = null, CancellationToken cancellationToken = default)
+    {
+        var embeddings = await InnerGenerator.GenerateAsync(values, options, cancellationToken).ConfigureAwait(false);
+
+        foreach (var e in embeddings)
+        {
+            if (!e.Normalized)
```

Comment:
Given that calculating the norm is a relatively speaking a cheap operation, couldn't we just decide to normalize on the basis of how close to 1 the value is?

Reply:
(I assume that this middleware wouldn't be inserted unless the inner generator doesn't normalize embeddings or isn't reliable in how it normalizes embeddings and in either case this check seems redundant).

Reply:
Not necessarily, but this middleware could choose to ignore it if we keep it. Perhaps it makes more sense to advertise on the generator itself rather than individual vectors.

Reply:
> Perhaps it makes more sense to advertise on the generator itself rather than individual vectors.

So remove Embedding<T>.Normalized and add EmbeddingGeneratorMetadata.CreatesNormalizedVectors, or something like that? It wouldn't make sense for some other kinds of embeddings, but in such cases it could just be false.

@SteveSandersonMS, opinions?

Reply:
Yep. Either that or `IsNormalizingGenerator` sounds reasonable to me.

Reply:
I thought the value of having this flag is so you can use it to make operations cheaper, e.g.:

```cs
var similarity = embedding1.Normalized && embedding2.Normalized
    ? TensorPrimitives.DotProduct(embedding1.Vector.Span, embedding2.Vector.Span)
    : TensorPrimitives.CosineSimilarity(embedding1.Vector.Span, embedding2.Vector.Span);
```

Having a flag on the embedding generator doesn't help with this unless the instance holds a reference to its original generator (which I don't think it does).

If it's not for this purpose, what purpose is it for?

The other main way to do something here would be to have a property like:

```
private float? _magnitude;

public float Magnitude {
    get {
        if (!magnitude.HasValue) {
            _magnitude = TensorPrimitives.WhateverComputesTheMagnitudeOrNormOrLength(Vector.Span);
        }

        return _magnitude.Value;
    }
    set => _magnitude = Value;
}
```

... so that people computing similarity can always do:

```cs
var similarity = TensorPrimitives.DotProduct(embedding1.Vector.Span, embedding2.Vector.Span)
    / (embedding1.Magnitude * embedding2.Magnitude);
```

... and the optimization takes effect if the generator chooses to assign `Magnitude = 1` because it knows the provider always does this.

Reply:
I think Eirik's suggestion is that the consumer just knows what generator they're working with / assumes things are always normalized, and if they're not, this can be used to make them so. Eirik?

Personally, I like having the metadata on the embedding instance.

Reply:
We haven't been super clear on what scenarios are improved by this change, which is making it quite hard to pick between different design possibilities. Perhaps we could frame it as two separate potential benefits:

 1. Guaranteeing that the vector is normalized
 2. Indicating whether it is normalized (either per-embedding or per-generator)

Somebody would want (1) if they have an embedding generator that actually produces non-normalized output but they want it to be normalized (e.g., so they can use dot product instead of cosine calculation elsewhere). This isn't a scenario we know exists because the generators we care about do produce normalized output. However, if someone is in this scenario, it's quite easy for them to add a `UseNormalization` step that normalizes unconditionally, or we can ship one in the library that we don't suggest you use with OpenAI/etc but it's there if you need it for some reason.

Somebody would want (2) if they want to optimize more aggressively, skipping the normalization step if it's prenormalized, or dynamically switching between dot product and cosine calculation for similarity on a per-embedding-pair basis.

 * For app developers, this is unlikely to be the case, since they likely know their vectors are prenormalized by the generator, and if not, they could put in an unconditional normalization step when they register the generator
 * For library authors, this is more plausible since they don't know or control which generator is used. However it's really niche to need this level of optimization instead of just computing the norm when they need it, or using `TensorPrimitives.CosineSimilarity` if it's a similarity calculation.

As for the risks of confusion/misuse:

 1. Shipping an unconditional `UseNormalization` is basically zero risk. People likely won't use it except in the rare case they need to. Even though we don't have a scenario where it's needed today, it will 100% produce the right output in all cases if used, so is not harmful.
 2. Keeping track of normalization status is a little more likely to get into confusion:
    * If it's at the embedding level, the flag could be wrong if someone misunderstands (e.g., they set `Normalized` to true because they *want* it to be normalized, not because they think it already is)
    * If it's at the generator level, the flag could be wrong because the truth might depend on which model/deployment is used, and the client library might not be able to tell from that string alone. Worst case is, e.g., OpenAI always produces normalized output today so it's hardcoded to `true` in all cases, but then one day they release a new embedding model that doesn't prenormalize.

Altogether I'm not hugely concerned what we do here, as it's all likely to be fine. If it were me choosing, I'd probably ship an unconditional `UseNormalization` and leave open the option to add further "is prenormalized" metadata in the future (which would then cause UseNormalization to no-op). But if you want to add the metadata now and/or track it on the embedding instances, it's all fine with me.

Reply:
I don't necessarily mind having metadata on the embedding instance itself, the only concern is that the metadata could be lying. I tend to agree with @SteveSandersonMS's takeaway that we should just ship an unconditional normalizer.

> Having a flag on the embedding generator doesn't help with this unless the instance holds a reference to its original generator (which I don't think it does).

I think this suggests we might want to expose magnitude/normalization functionality on the embedding type itself:
```C#
public static class EmbeddingExtensions
{
    public static bool IsNormalized(this Embedding<T> embedding) where T : IRootFunctions<T> { }
    public static T Magnitude<T>(this Embedding<T> embedding) where T : IRootFunctions<T> {  }
    public static void Normalize<T>(this Embedding<T> embedding) where T : IRootFunctions<T> { }
}
```
A normalizing middleware can be trivially built on top of this and as an added bonus it avoids the ambiguities of exposing such APIs in non-floating point embeddings.

Reply:
> I think this suggests we might want to expose magnitude/normalization functionality on the embedding type itself:

I guess I just don't see the benefit of this. It's a one liner with TensorPrimitives.

I'll just remove the metadata and make the normalization unconditional.

Reply:
Perhaps there's benefit in defining the appropriate tolerances for declaring something as being normalized and handling zero vectors correctly.

Reply:
> and handling zero vectors correctly

We obviously disagree here. I believe TP handles zero vectors correctly. And if someone wants tolerances, that's something I want them defining, not us.

Reply:
> I believe TP handles zero vectors correctly. 

Does TP have a normalization routine?

Reply:
That was a general statement. It has a Norm method and it has a Divide method.